### PR TITLE
Instructor: remind particular students: show sorted student list #8630

### LINKED ISSUE #8630: Instructor: remind particular students: show sorted student list
**Description of feature/enhancement**

Current:
screen shot 2018-03-12 at 5 12 02 pm
screen shot 2018-03-12 at 5 12 15 pm

Suggestion:
In the modal opened up by `Remind particular students` menu option, by default show student list sorted by `Team` > `Student Name` for each of `not responded` and `responded` categories.

**Justification**

Sorted student list helps instructors to locate specific students faster.

### PR DESCRIPTION
**Pull request by mbi90**
Fixes #8630

**Outline of Solution**
Sorts the lists of students (noResponse and studentsWhoResponded) by teams then by username.
@Reviewers: I'm not sure about the placement of the code. If you would prefere the class to stay as a model (logic free) then I should move the logic somewhere else.

### CODE REVIEW

**Comment by mbi90** [comment]
@bqnguyen94 a suggestion to fix #8630 - this is my first PR so feel free to require changes

File: src/main/java/teammates/ui/pagedata/InstructorFeedbackRemindParticularStudentsPageData.java [hunk 1]
```diff
@@ -28,4 +32,24 @@ public String getFsName() {
     public FeedbackSessionResponseStatus getResponseStatus() {
         return responseStatus;
     }
+
+    /**
+     * sorts the students in the responseStatus (both who responded and with no response).
+     * The sort is by Team then by student name.
+     */
+    private void sortStudentNamesListsInResponseStatus() {
```
Comment (bqnguyen94) [inline src/main/java/teammates/ui/pagedata/InstructorFeedbackRemindParticularStudentsPageData.java]: If you look into `teammates.common.datatransfer.FeedbackSessionResponseStatus`, you will see that the comparison functions are already there. No need to reinvent the wheel here.

**Comment by mbi90** [comment]
@bqnguyen94 thanks for your advices. I didn't know that jsp automatically uses the getter.

After making these last changes I find that the methods `FeedbackSessionResponseStatus::getNoResponse()` and `FeedbackSessionResponseStatus::getStudentsWhoDidNotRespondToAnyQuestion()` are identical.
Would you prefere removing one of them?

**Comment by bqnguyen94** [comment]
I agree, after adding `sort` the 2 functions are identical. However, I feel removing identical functions is out of scope for this PR. You can open a new issue for it.

**Comment by mbi90** [comment]
Ready for reviewing

File: src/main/java/teammates/ui/pagedata/InstructorFeedbackRemindParticularStudentsPageData.java [hunk 2]
```diff
@@ -9,8 +9,8 @@
     private String fsName;
 
     public InstructorFeedbackRemindParticularStudentsPageData(
-                AccountAttributes account, String sessionToken, FeedbackSessionResponseStatus responseStatus,
-                String courseId, String fsName) {
+            AccountAttributes account, String sessionToken, FeedbackSessionResponseStatus responseStatus,
+            String courseId, String fsName) {
```
Comment (bqnguyen94) [inline src/main/java/teammates/ui/pagedata/InstructorFeedbackRemindParticularStudentsPageData.java]: These changes are not relevant right? Can you undo them? :P

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java [hunk 3]
```diff
@@ -68,15 +68,19 @@ public FeedbackSessionResponseStatus() {
 
     /**
      * Returns list of students who did not respond to the feedback session.
```
Comment (bqnguyen94) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java]: Final touch, can you remove the period at the end?

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java [hunk 4]
```diff
@@ -68,15 +68,19 @@ public FeedbackSessionResponseStatus() {
...
+     * sorted by teamName > studentNamelist.
      */
     public List<String> getNoResponse() {
+        noResponse.sort(compareByTeamNameStudentName);
         return noResponse;
     }
 
     /**
      * Returns list of students who responded to the feedback session.
```
Comment (bqnguyen94) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java]: Same here.

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java (hunk 3 above)
Comment (tshradheya) [reply src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java]: I think the period is necessary for JavaDocs.

If removed checkstyle gives an error.

Comment (wkurniawan07) [reply src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java]: @tshradheya take a look at the next line. @bqnguyen94 is giving a legitimate suggestion here.

Comment (tshradheya) [reply src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java]: Ohh sorry 😅 My bad

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java (hunk 4 above)
Comment (mbi90) [reply src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java]: This is actually part of the fix. we need the list to be sorted :)

**Comment by mbi90** [comment]
@bqnguyen94 thanks for your reviews. One last time :) 👍

**Comment by mbi90** [comment]
pls review

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java (hunk 3 above)
Comment (bqnguyen94) [reply src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java]: @mbi90 you haven't update the code here?

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java (hunk 4 above)
Comment (bqnguyen94) [reply src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java]: @mbi90 and here?

Comment (mbi90) [reply src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java]: @bqnguyen94 this and line 83 are actually the fix!

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java (hunk 3 above)
Comment (mbi90) [reply src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java]: I corrected the place of the period. I tried removing the period completely but that caused a failure (at gradle lint)

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java [hunk 5]
```diff
@@ -58,8 +58,8 @@ public FeedbackSessionResponseStatus() {
     }
 
     /**
-     * Returns list of students who did not respond to the feedback session
-     * sorted by teamName > studentNamelist.
+     * Returns list of students who did not respond to the feedback session.
+     * sorted by teamName > studentNamelist
```
Comment (bqnguyen94) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java]: Period should be at the end of sentence.

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java (hunk 3 above)
Comment (bqnguyen94) [reply src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java]: Hmm that's strange. What's the error message?

**Comment by bqnguyen94** [comment]
@whipermr5 this LGTM

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java [hunk 6]
```diff
@@ -67,16 +67,20 @@ public FeedbackSessionResponseStatus() {
     }
 
     /**
-     * Returns list of students who did not respond to the feedback session.
+     * Returns list of students who did not respond to the feedback session
+     * sorted by teamName > studentNamelist.
      */
     public List<String> getNoResponse() {
+        noResponse.sort(compareByTeamNameStudentName);
```
Comment (whipermr5) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java]: Now this does exactly the same as `getStudentsWhoDidNotRespondToAnyQuestion`. Instead of modifying this getter, rename `getStudentsWhoDidNotRespondToAnyQuestion` -> `getStudentsWhoDidNotRespondSorted`. Also, this is a good chance to rename `noResponse` -> `studentsWhoDidNotRespond`.

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java [hunk 7]
```diff
@@ -67,16 +67,20 @@ public FeedbackSessionResponseStatus() {
...
         return noResponse;
     }
 
     /**
-     * Returns list of students who responded to the feedback session.
+     * Returns list of students who responded to the feedback session
+     * sorted by teamName > studentNamelist.
      */
     public List<String> getStudentsWhoResponded() {
+        studentsWhoResponded.sort(compareByTeamNameStudentName);
```
Comment (whipermr5) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java]: Similar to above - revert the changes here and add a new method `getStudentsWhoRespondedSorted`, then use that where we need the sorted list.

**Comment by mbi90** [comment]
@whipermr5 done :)

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java [hunk 8]
```diff
@@ -61,16 +61,25 @@ public FeedbackSessionResponseStatus() {
...
+        studentsWhoDidNotRespond.sort(compareByTeamNameStudentName);
+        return studentsWhoDidNotRespond;
     }
 
     /**
      * Returns list of students who did not respond to the feedback session.
      */
-    public List<String> getNoResponse() {
-        return noResponse;
+    public List<String> getstudentsWhoDidNotRespond() {
```
Comment (whipermr5) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResponseStatus.java]: Capital S

File: src/main/webapp/jsp/instructorFeedbackAjaxRemindParticularStudentsModal.jsp [hunk 9]
```diff
@@ -26,7 +26,7 @@
       <span class="icon-sort unsorted"></span>
     </th>
   </tr>
-  <c:forEach items="${data.responseStatus.noResponse}" var="userToRemindEmail">
+  <c:forEach items="${data.responseStatus.getStudentsWhoDidNotRespondSorted()}" var="userToRemindEmail">
```
Comment (whipermr5) [inline src/main/webapp/jsp/instructorFeedbackAjaxRemindParticularStudentsModal.jsp]: You'll find that changing this to `${data.responseStatus.studentsWhoDidNotRespondSorted}` will work out of the box (it automatically calls `getStudentsWhoDidNotRespondSorted()`).

File: src/main/webapp/jsp/instructorFeedbackAjaxRemindParticularStudentsModal.jsp [hunk 10]
```diff
@@ -52,7 +52,7 @@
       </td>
     </tr>
   </c:forEach>
-  <c:forEach items="${data.responseStatus.studentsWhoResponded}" var="userToRemindEmail">
+  <c:forEach items="${data.responseStatus.getStudentsWhoRespondedSorted()}" var="userToRemindEmail">
```
Comment (whipermr5) [inline src/main/webapp/jsp/instructorFeedbackAjaxRemindParticularStudentsModal.jsp]: Same here: `${data.responseStatus.studentsWhoRespondedSorted}`.

**Comment by mbi90** [comment]
@whipermr5 should be ready at last! 👍

**Comment by whipermr5** [comment]
@mbi90 Thanks for your contribution! May we have your full name for inclusion into the contributors' list? :)
